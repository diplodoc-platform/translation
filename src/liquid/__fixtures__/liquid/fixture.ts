export type Fixtures = Array<Fixture>;

export type Fixture = {
    section: string;
    content: string;
    number: number;
};
