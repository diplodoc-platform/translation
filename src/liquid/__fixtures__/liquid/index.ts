export * from './fixture';
export * from './text';
export * from './condition';
export * from './functions';
export * from './filter';
export * from './variable';
export * from './loop';
export * as default from './index';
