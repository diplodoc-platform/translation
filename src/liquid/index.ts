export {Tokenizer} from './tokenizer';
export * from './token';
export * from './renderer';
export * from './index';
