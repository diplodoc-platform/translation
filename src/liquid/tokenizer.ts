import {Token, TokenType} from './token';

export type Configuration = {
    specification?: Specification;
};

export type Specification = Array<SpecificationEntry>;
export type SpecificationEntry = [RegExp, TokenType];

const specification_: Specification = [
    // Conditions
    // If statement
    [/^\{%[^\S\r\n]*if[^%}]+?[^\S\r\n]*%}/, 'If'],
    // Else statement
    [/^\{%[^\S\r\n]*else[^\S\r\n]*%\}/, 'Else'],
    // EndIf statement
    [/^\{%[^\S\r\n]*endif[^\S\r\n]*%\}/, 'EndIf'],
    // Function
    [/^\{\{[^\S\r\n]*[\w.-]+?\(.*?\)[^\S\r\n]*\}\}/, 'Function'],
    // Filter
    [/^\{\{[^\S\r\n]*[\w.]+[^\S\r\n]*\|[^\S\r\n]*\w+[^\S\r\n]*\}\}/, 'Filter'],
    // Variable
    [/^\{\{[^\S\r\n]*[\w.]+[^\S\r\n]*\}\}/, 'Variable'],
    // ForInLoop
    [/^\{%[^\S\r\n]*for[^\S\r\n]+[\w.]+[^\S\r\n]+in[^\S\r\n]+[\w.]+[^\S\r\n]*%\}/, 'ForInLoop'],
    // EndForInLoop
    [/^\{%[^\S\r\n]*endfor[^\S\r\n]*%\}/, 'EndForInLoop'],
    // Space
    [/^[^\S\r\n]+/, 'Space'],
    // Newline
    [/^[\r\n]+/, 'Newline'],
    // Text
    // without grabbing liquid/variable syntax
    [/^[^\s]+(?={{|{%)/, 'Text'],
    // plain text
    [/^[^\s]+/, 'Text'],
];

export type TokenizerGenerator = Generator<Token | null, void, Token | undefined>;

class Tokenizer implements TokenizerGenerator {
    private input: string;
    private cursor: number;
    private specification: Specification;

    constructor(input: string, configuration: Configuration = {}) {
        this.input = input;
        this.cursor = 0;

        const {specification = specification_} = configuration;
        this.specification = specification;
    }

    tokenize(this: TokenizerGenerator & this): Token[] {
        this.cursor = 0;

        return Array.from<Token>(this);
    }

    *[Symbol.iterator](this: TokenizerGenerator & this) {
        let value, done;

        do {
            ({value, done} = this.next() ?? {});
            // eslint-disable-next-line no-eq-null, eqeqeq
            if (value == null) {
                return;
            }

            yield value;
        } while (!done);
    }

    next(): IteratorResult<Token | null> {
        const token = this.match();
        if (!token) {
            return {value: null, done: true};
        }

        this.cursor += token.value.length;

        return {value: token, done: this.done()};
    }

    return() {
        return this.next();
    }

    throw(): IteratorResult<Token> {
        return {value: null, done: true};
    }

    private match(this: Tokenizer) {
        const left = this.input.slice(this.cursor);

        let value;

        for (const [regexp, type] of this.specification) {
            [value] = regexp.exec(left) ?? [];
            // eslint-disable-next-line eqeqeq, no-eq-null
            if (value == null) {
                continue;
            }

            return {
                type,
                value,
            };
        }

        return null;
    }

    private done() {
        return this.cursor === this.input.length;
    }
}

export {Tokenizer};
export default {Tokenizer};
