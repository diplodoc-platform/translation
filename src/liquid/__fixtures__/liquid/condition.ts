import {Fixtures} from './fixture';

const condition: Fixtures = [
    {
        section: 'conditions',
        content: "Person is: {% if personType == 'cool' %} cool {% else %} not cool.",
        number: 1,
    },
    {
        section: 'conditions',
        content:
            '{%if var == "a"%} A {%else%} B was the point we examined. Now examine let\'s examine others.',
        number: 2,
    },
    {
        section: 'conditions',
        content: `{% if var == "a" %}

Content when var is equal a

{% else %}

Content when var is not equal a

{% endif %}

`,
        number: 3,
    },
];

export {condition};
export default {condition};
