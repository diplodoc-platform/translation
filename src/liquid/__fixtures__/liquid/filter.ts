import {Fixtures} from './fixture';

const filter: Fixtures = [
    {
        section: 'filter',
        content: 'Hello P{{user.name | capitalize}}!',
        number: 1,
    },
    {
        section: 'filter',
        content: 'We have {{user.name | length}} users.',
        number: 2,
    },
    {
        section: 'filter',
        content: 'Users amount: {{ user.name | length }}.',
        number: 3,
    },
    {
        section: 'filter',
        content: 'Users amount: {{ user_name-cool | length }}.',
        number: 4,
    },
];

export {filter};
export default {filter};
