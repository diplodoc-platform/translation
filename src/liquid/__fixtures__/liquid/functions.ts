import {Fixtures} from './fixture';

const functions: Fixtures = [
    {
        section: 'functions',
        content: 'Hello P{{ user.name.slice(1, 2) }}vel!',
        number: 1,
    },
    {
        section: 'functions',
        content: 'Sentence {{user.name.slice(1)}} content after.',
        number: 2,
    },
    {
        section: 'functions',
        content: '{{user.slice(1)}}.',
        number: 3,
    },
    {
        section: 'functions',
        content: '{{user_name-cool.slice(1)}}.',
        number: 4,
    },
];

export {functions};
export default {functions};
