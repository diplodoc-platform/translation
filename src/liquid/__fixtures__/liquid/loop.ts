import {Fixtures} from './fixture';

const loop: Fixtures = [
    {
        section: 'loop',
        content: `{% for x in xs %} x {% endfor %}`,
        number: 1,
    },
    {
        section: 'loop',
        content: `{%for x in xs%} x {%endfor%}`,
        number: 2,
    },
    {
        section: 'loop',
        content: `{%for x.a_b in xs.ys%} x {%endfor%}`,
        number: 3,
    },
    {
        section: 'loop',
        content: `\
{%for x.a_b in xs.ys%}

{{a_b.c}}

{%endfor%}`,
        number: 4,
    },
];

export {loop};
export default {loop};
