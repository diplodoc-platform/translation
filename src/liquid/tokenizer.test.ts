/* eslint-disable @typescript-eslint/no-loop-func */
import {Tokenizer} from './tokenizer';
import {condition, filter, functions, loop, text, variable} from './__fixtures__/liquid';

const fixtures = [...text, ...condition, ...functions, ...filter, ...variable, ...loop];

describe('liquid tokenizer smoke', () => {
    it('works', () => {
        const tokenizer = new Tokenizer('');
        tokenizer.tokenize();
    });
});

describe('liquid tokenizer syntax', () => {
    for (const {section, number, content} of fixtures) {
        test(`${section}-${number}`, () => {
            const tokenizer = new Tokenizer(content);
            const tokens = [];
            for (const token of tokenizer) {
                tokens.push(token);
            }
            expect(tokens).toMatchSnapshot();
        });
    }
});
