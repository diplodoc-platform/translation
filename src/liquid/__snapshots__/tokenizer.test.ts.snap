// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`liquid tokenizer syntax conditions-1 1`] = `
Array [
  Object {
    "type": "Text",
    "value": "Person",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "is:",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "If",
    "value": "{% if personType == 'cool' %}",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "cool",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Else",
    "value": "{% else %}",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "not",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "cool.",
  },
]
`;

exports[`liquid tokenizer syntax conditions-2 1`] = `
Array [
  Object {
    "type": "If",
    "value": "{%if var == \\"a\\"%}",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "A",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Else",
    "value": "{%else%}",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "B",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "was",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "the",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "point",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "we",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "examined.",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "Now",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "examine",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "let's",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "examine",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "others.",
  },
]
`;

exports[`liquid tokenizer syntax conditions-3 1`] = `
Array [
  Object {
    "type": "If",
    "value": "{% if var == \\"a\\" %}",
  },
  Object {
    "type": "Newline",
    "value": "

",
  },
  Object {
    "type": "Text",
    "value": "Content",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "when",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "var",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "is",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "equal",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "a",
  },
  Object {
    "type": "Newline",
    "value": "

",
  },
  Object {
    "type": "Else",
    "value": "{% else %}",
  },
  Object {
    "type": "Newline",
    "value": "

",
  },
  Object {
    "type": "Text",
    "value": "Content",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "when",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "var",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "is",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "not",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "equal",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "a",
  },
  Object {
    "type": "Newline",
    "value": "

",
  },
  Object {
    "type": "EndIf",
    "value": "{% endif %}",
  },
  Object {
    "type": "Newline",
    "value": "

",
  },
]
`;

exports[`liquid tokenizer syntax filter-1 1`] = `
Array [
  Object {
    "type": "Text",
    "value": "Hello",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "P",
  },
  Object {
    "type": "Filter",
    "value": "{{user.name | capitalize}}",
  },
  Object {
    "type": "Text",
    "value": "!",
  },
]
`;

exports[`liquid tokenizer syntax filter-2 1`] = `
Array [
  Object {
    "type": "Text",
    "value": "We",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "have",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Filter",
    "value": "{{user.name | length}}",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "users.",
  },
]
`;

exports[`liquid tokenizer syntax filter-3 1`] = `
Array [
  Object {
    "type": "Text",
    "value": "Users",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "amount:",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Filter",
    "value": "{{ user.name | length }}",
  },
  Object {
    "type": "Text",
    "value": ".",
  },
]
`;

exports[`liquid tokenizer syntax filter-4 1`] = `
Array [
  Object {
    "type": "Text",
    "value": "Users",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "amount:",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Filter",
    "value": "{{ user_name-cool | length }}",
  },
  Object {
    "type": "Text",
    "value": ".",
  },
]
`;

exports[`liquid tokenizer syntax functions-1 1`] = `
Array [
  Object {
    "type": "Text",
    "value": "Hello",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "P",
  },
  Object {
    "type": "Function",
    "value": "{{ user.name.slice(1, 2) }}",
  },
  Object {
    "type": "Text",
    "value": "vel!",
  },
]
`;

exports[`liquid tokenizer syntax functions-2 1`] = `
Array [
  Object {
    "type": "Text",
    "value": "Sentence",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Function",
    "value": "{{user.name.slice(1)}}",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "content",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "after.",
  },
]
`;

exports[`liquid tokenizer syntax functions-3 1`] = `
Array [
  Object {
    "type": "Function",
    "value": "{{user.slice(1)}}",
  },
  Object {
    "type": "Text",
    "value": ".",
  },
]
`;

exports[`liquid tokenizer syntax functions-4 1`] = `
Array [
  Object {
    "type": "Function",
    "value": "{{user_name-cool.slice(1)}}",
  },
  Object {
    "type": "Text",
    "value": ".",
  },
]
`;

exports[`liquid tokenizer syntax loop-1 1`] = `
Array [
  Object {
    "type": "ForInLoop",
    "value": "{% for x in xs %}",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "x",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "EndForInLoop",
    "value": "{% endfor %}",
  },
]
`;

exports[`liquid tokenizer syntax loop-2 1`] = `
Array [
  Object {
    "type": "ForInLoop",
    "value": "{%for x in xs%}",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "x",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "EndForInLoop",
    "value": "{%endfor%}",
  },
]
`;

exports[`liquid tokenizer syntax text-1 1`] = `
Array [
  Object {
    "type": "Text",
    "value": "Sentence",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "with",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "plain",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "text.",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "Another",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "plain",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "text",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "sentence.",
  },
]
`;

exports[`liquid tokenizer syntax text-2 1`] = `
Array [
  Object {
    "type": "Text",
    "value": "Sentence.",
  },
  Object {
    "type": "Space",
    "value": "    ",
  },
  Object {
    "type": "Text",
    "value": "Sentence",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "after",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "4",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "spaces.",
  },
]
`;

exports[`liquid tokenizer syntax variable-1 1`] = `
Array [
  Object {
    "type": "Text",
    "value": "Hello",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Variable",
    "value": "{{user.name}}",
  },
  Object {
    "type": "Text",
    "value": "!",
  },
]
`;

exports[`liquid tokenizer syntax variable-2 1`] = `
Array [
  Object {
    "type": "Text",
    "value": "We",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "have",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Variable",
    "value": "{{ user.name }}",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "users.",
  },
]
`;

exports[`liquid tokenizer syntax variable-3 1`] = `
Array [
  Object {
    "type": "Text",
    "value": "Users",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Text",
    "value": "amount:",
  },
  Object {
    "type": "Space",
    "value": " ",
  },
  Object {
    "type": "Variable",
    "value": "{{user.name}}",
  },
  Object {
    "type": "Text",
    "value": ".",
  },
]
`;
